File: android/src/main/kotlin/in/agnostech/flutter_stripe_terminal/FlutterStripeTerminalEventHandler.kt
package `in`.agnostech.flutter_stripe_terminal

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.stripe.stripeterminal.Terminal
import com.stripe.stripeterminal.external.callable.*
import com.stripe.stripeterminal.external.models.*
import com.stripe.stripeterminal.log.LogLevel
import io.flutter.plugin.common.EventChannel

class FlutterStripeTerminalEventHandler(private val context: Context): EventChannel.StreamHandler, TerminalListener, DiscoveryListener, BluetoothReaderListener {

    private lateinit var eventSink: EventChannel.EventSink

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        Log.d("STRIPE TERMINAL", "event listener called")
        this.eventSink = events
        val logLevel = LogLevel.VERBOSE
        val tokenProvider = TokenProvider()
        if (!Terminal.isInitialized()) {
            Terminal.initTerminal(context, logLevel, tokenProvider, this)
        }
    }

    fun getDiscoveryListener(): DiscoveryListener {
        return this
    }

    fun getBluetoothReaderListener(): BluetoothReaderListener {
        return this
    }

    override fun onCancel(arguments: Any?) {

    }

    override fun onConnectionStatusChange(status: ConnectionStatus) {
        super.onConnectionStatusChange(status)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerConnectionStatus" to status.name
            ))
        }
    }

    override fun onPaymentStatusChange(status: PaymentStatus) {
        super.onPaymentStatusChange(status)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerPaymentStatus" to status.name
            ))
        }
    }

    override fun onUnexpectedReaderDisconnect(reader: Reader) {
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerConnectionStatus" to "DISCONNECTED"
            ))
        }
    }

    override fun onUpdateDiscoveredReaders(readers: List<Reader>) {
        FlutterStripeTerminal.availableReadersList = readers

        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "deviceList" to readers.map {
                    mapOf(
                        "id" to it.serialNumber,
                        "deviceName" to it.deviceType.name
                    )
                }
            ))
        }
    }

    override fun onFinishInstallingUpdate(update: ReaderSoftwareUpdate?, e: TerminalException?) {
        super.onFinishInstallingUpdate(update, e)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerUpdateStatus" to "FINISHED_UPDATE_INSTALLATION"
            ))
        }
    }

    override fun onReportAvailableUpdate(update: ReaderSoftwareUpdate) {
        super.onReportAvailableUpdate(update)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerUpdateStatus" to "UPDATE_AVAILABLE"
            ))
        }
    }

    override fun onReportLowBatteryWarning() {
        super.onReportLowBatteryWarning()
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerEvent" to "LOW_BATTERY"
            ))
        }
    }

    override fun onReportReaderEvent(event: ReaderEvent) {
        super.onReportReaderEvent(event)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerEvent" to event.name
            ))
        }
    }

    override fun onReportReaderSoftwareUpdateProgress(progress: Float) {
        super.onReportReaderSoftwareUpdateProgress(progress)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerUpdateStatus" to "SOFTWARE_UPDATE_IN_PROGRESS"
            ))
        }
    }

    override fun onRequestReaderDisplayMessage(message: ReaderDisplayMessage) {
        super.onRequestReaderDisplayMessage(message)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerEvent" to message.name
            ))
        }
    }

    override fun onRequestReaderInput(options: ReaderInputOptions) {
        super.onRequestReaderInput(options)
        Log.d("READER INPUT REQUEST", options.toString())
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerEvent" to options
            ))
        }
    }

    override fun onStartInstallingUpdate(update: ReaderSoftwareUpdate, cancelable: Cancelable?) {
        super.onStartInstallingUpdate(update, cancelable)
        Handler(Looper.getMainLooper()).post {
            eventSink.success(mapOf(
                "readerUpdateStatus" to "STARTING_UPDATE_INSTALLATION"
            ))
        }
    }
}

File: ios/Classes/FlutterStripeTerminalEventHandler.swift
//
//  FlutterStripeTerminalEventHandler.swift
//  flutter_stripe_terminal
//
//  Created by Vishal Dubey on 03/12/21.
//

import Foundation
import Flutter
import StripeTerminal

class FlutterStripeTerminalEventHandler: NSObject, FlutterStreamHandler, DiscoveryDelegate, TerminalDelegate, BluetoothReaderDelegate {
    
    static let shared = FlutterStripeTerminalEventHandler()
    var eventSink: FlutterEventSink?
    
    func reader(_ reader: Reader, didReportAvailableUpdate update: ReaderSoftwareUpdate) {
        eventSink!([
            "readerUpdateStatus": "UPDATE_AVAILABLE"
        ])
    }
    
    func reader(_ reader: Reader, didStartInstallingUpdate update: ReaderSoftwareUpdate, cancelable: Cancelable?) {
        eventSink!([
            "readerUpdateStatus": "STARTING_UPDATE_INSTALLATION"
        ])
    }
    
    func reader(_ reader: Reader, didReportReaderSoftwareUpdateProgress progress: Float) {
        eventSink!([
            "readerUpdateStatus": "SOFTWARE_UPDATE_IN_PROGRESS"
        ])
    }
    
    func reader(_ reader: Reader, didFinishInstallingUpdate update: ReaderSoftwareUpdate?, error: Error?) {
        eventSink!([
            "readerUpdateStatus": "FINISHED_UPDATE_INSTALLATION"
        ])
    }
    
    func reader(_ reader: Reader, didRequestReaderInput inputOptions: ReaderInputOptions = []) {
        eventSink!([
            "readerInputEvent": Terminal.stringFromReaderInputOptions(inputOptions)
        ])
    }
    
    func reader(_ reader: Reader, didRequestReaderDisplayMessage displayMessage: ReaderDisplayMessage) {
        eventSink!([
            "readerEvent": Terminal.stringFromReaderDisplayMessage(displayMessage)
        ])
    }
    
    func reader(_ reader: Reader, didReportBatteryLevel batteryLevel: Float, status: BatteryStatus, isCharging: Bool) {
        eventSink!([
            "readerEvent": "LOW_BATTERY"
        ])
    }
    
    func reader(_ reader: Reader, didReportReaderEvent event: ReaderEvent, info: [AnyHashable : Any]?) {
        eventSink!([
            "readerEvent": Terminal.stringFromReaderEvent(event)
        ])
    }
    
    func terminal(_ terminal: Terminal, didChangePaymentStatus status: PaymentStatus) {
        eventSink!([
            "readerPaymentStatus": Terminal.stringFromPaymentStatus(status)
        ])
    }
    
    func terminal(_ terminal: Terminal, didChangeConnectionStatus status: ConnectionStatus) {
        eventSink!([
            "readerConnectionStatus": Terminal.stringFromConnectionStatus(status)
        ])
    }
    
    func terminal(_ terminal: Terminal, didReportUnexpectedReaderDisconnect reader: Reader) {
        eventSink!([
            "readerConnectionStatus": "DISCONNECTED"
        ])
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        Terminal.setTokenProvider(APIClient.shared)
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    func terminal(_ terminal: Terminal, didUpdateDiscoveredReaders readers: [Reader]) {
        FlutterStripeTerminal.shared.availableReaders = readers
        eventSink!([
            "deviceList": readers.map{
                reader in
                return [
                    "serialNumber": reader.serialNumber,
                    "deviceName": Terminal.stringFromDeviceType(reader.deviceType)
                ]
            }
        ])
    }
    
}
